OpenRA.Mods.Cnc/Traits/OnRails.cs

```diff
@@ -0,0 +1,219 @@
+			return new ReadOnlyDictionary<CPos, SubCell>(occupied);
+		}
+
+		bool IOccupySpaceInfo.SharesCell { get { return false; } }
+
+		// Used to determine if actor can spawn
+		public bool CanEnterCell(World world, Actor self, CPos cell, Actor ignoreActor = null, bool checkTransientActors = false)
+		{
+			if (!world.Map.Contains(cell))
+				return false;
+
+			return true;
+		}
+	}
+
+	public class OnRails : ITick, ISync, IFacing, IPositionable, IMove, IDeathActorInitModifier,
+		INotifyCreated, INotifyAddedToWorld, INotifyRemovedFromWorld, IActorPreviewInitModifier
+	{
+		public readonly OnRailsInfo Info;
+		readonly Actor self;
+
+		IEnumerable<int> speedModifiers;
+
+		[Sync] public int Facing { get; set; }
+		[Sync] public WPos CenterPosition { get; private set; }
+		public CPos TopLeft { get { return self.World.Map.CellContaining(CenterPosition); } }
+
+		// Isn't used anyway
+		public int TurnSpeed { get { return 255; } }
+
+		CPos cachedLocation;
+
+		public OnRails(ActorInitializer init, OnRailsInfo info)
+		{
+			Info = info;
+			self = init.Self;
+
+			if (init.Contains<LocationInit>())
+				SetPosition(self, init.Get<LocationInit, CPos>());
+
+			if (init.Contains<CenterPositionInit>())
+				SetPosition(self, init.Get<CenterPositionInit, WPos>());
+
+			Facing = init.Contains<FacingInit>() ? init.Get<FacingInit, int>() : Info.GetInitialFacing();
+
+			// Prevent mappers from setting bogus facings
+			if (Facing != 64 && Facing != 192)
+				Facing = Facing > 127 ? 192 : 64;
+		}
+
+		void INotifyCreated.Created(Actor self)
+		{
+			speedModifiers = self.TraitsImplementing<ISpeedModifier>().ToArray().Select(sm => sm.GetSpeedModifier());
+			cachedLocation = self.Location;
+		}
+
+		void INotifyAddedToWorld.AddedToWorld(Actor self)
+		{
+			self.World.AddToMaps(self, this);
+		}
+
+		void INotifyRemovedFromWorld.RemovedFromWorld(Actor self)
+		{
+			self.World.RemoveFromMaps(self, this);
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (cachedLocation != self.Location)
+			{
+				// If the actor just left the map, switch facing
+				if (!self.World.Map.Contains(self.Location))
+					Turn();
+			}
+
+			cachedLocation = self.Location;
+
+			SetVisualPosition(self, self.CenterPosition + MoveStep(Facing));
+		}
+
+		void Turn()
+		{
+			if (Facing == 64)
+				Facing = 192;
+			else
+				Facing = 64;
+		}
+
+		int MovementSpeed
+		{
+			get { return Util.ApplyPercentageModifiers(Info.Speed, speedModifiers); }
+		}
+
+		public IEnumerable<Pair<CPos, SubCell>> OccupiedCells() { return new[] { Pair.New(TopLeft, SubCell.FullCell) }; }
+
+		WVec MoveStep(int facing)
+		{
+			return MoveStep(MovementSpeed, facing);
+		}
+
+		WVec MoveStep(int speed, int facing)
+		{
+			var dir = new WVec(0, -1024, 0).Rotate(WRot.FromFacing(facing));
+			return speed * dir / 1024;
+		}
+
+		void IDeathActorInitModifier.ModifyDeathActorInit(Actor self, TypeDictionary init)
+		{
+			init.Add(new FacingInit(Facing));
+		}
+
+		#region Implement IPositionable
```

Comment:
`[15:46:16] <pchote> we implicitly prefer explicit interface implementations over regions`, which probably means you should remove the regions and implement those interfaces explicitly.